crates/adapters/src/format/avro/serializer.rs

```diff
@@ -0,0 +1,826 @@
+    }
+
+    fn unknown_field(field: &str, struct_name: String) -> Self {
+        AvroSerializerError::UnknownField {
+            field: field.to_string(),
+            struct_name,
+        }
+    }
+    fn wrong_number_of_fields(typ: &str, expected: usize, actual: usize) -> Self {
+        AvroSerializerError::WrongNumberOfFields {
+            typ: typ.to_string(),
+            expected,
+            actual,
+        }
+    }
+}
+
+impl std::error::Error for AvroSerializerError {}
+
+impl Display for AvroSerializerError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            AvroSerializerError::OutOfBounds { value, typ, schema } => {
+                write!(
+                    f,
+                    "value '{value}' of type '{typ}' is out of bounds of AVRO type {schema:?}"
+                )
+            }
+            AvroSerializerError::Incompatible { typ, schema } => {
+                write!(
+                    f,
+                    "value of type '{typ}' cannot be serialized using AVRO schema {schema:?}"
+                )
+            }
+            AvroSerializerError::MapKeyNotAString { key: Some(key) } => {
+                write!(f, "map key '{key}' is not a string")
+            }
+            AvroSerializerError::MapKeyNotAString { key: None } => {
+                write!(f, "map key is not a string")
+            }
+            AvroSerializerError::UnknownField { field, struct_name } => {
+                write!(
+                    f,
+                    "field '{field}' is not a member of struct '{struct_name}'"
+                )
+            }
+            AvroSerializerError::WrongNumberOfFields {
+                typ,
+                expected,
+                actual,
+            } => {
+                write!(
+                    f,
+                    "wrong number of fields in '{typ}' (expected: {expected}, provided: {actual})"
+                )
+            }
+            AvroSerializerError::Custom { error } => f.write_str(error),
+        }
+    }
+}
+
+impl serde::ser::Error for AvroSerializerError {
+    fn custom<T>(msg: T) -> Self
+    where
+        T: Display,
+    {
+        Self::Custom {
+            error: msg.to_string(),
+        }
+    }
+}
+
+pub struct StructSerializer<'a> {
+    schema: &'a RecordSchema,
+    fields: Vec<(String, AvroValue)>,
+}
+
+impl<'a> StructSerializer<'a> {
+    pub fn new(schema: &'a RecordSchema) -> Self {
+        Self {
+            schema,
+            fields: Vec::with_capacity(schema.fields.len()),
+        }
+    }
+}
+
+impl<'a> SerializeStruct for StructSerializer<'a> {
+    type Ok = AvroValue;
+    type Error = AvroSerializerError;
+
+    fn serialize_field<T: ?Sized>(
+        &mut self,
+        name: &'static str,
+        value: &T,
+    ) -> Result<(), Self::Error>
+    where
+        T: Serialize,
+    {
+        let field_schema_idx = *self.schema.lookup.get(name).ok_or_else(|| {
```

Comment:
could this lookup get cached?

Reply:
It's already a map indexed by name, not a lot to speed up here, without changing the whole design.